src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala

```diff
@@ -176,124 +176,194 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+          }
+        }
+        case (x: Cond, _) => Some(compileCond(x, z))
+        case (x: ModifierAtom, index) => {
+          val newIndexPrefix = s"${indexPrefix}${index}_"
+          val subqueryWhereStr = generateWhereClause(x.bodies, newIndexPrefix)
+          val subqueryFromStr  = generateFromClause(x.bodies, newIndexPrefix)
+          x.modifier match {
+            case ExistModifier(negated) => Some(s"${if (negated) "NOT " else ""}EXISTS (SELECT 1 FROM ${subqueryFromStr} WHERE ${subqueryWhereStr})")
+            case _ => None
+          }
+        }
+      }
+      conditions.mkString(" AND ")
+    }
+
+    def generateFromClause(bodies: List[Body], indexPrefix: String) : String = {
+      val atoms = bodies.zipWithIndex flatMap {
+        case (x:Atom,i) => Some(s"${x.name} R${indexPrefix}${i}")
+        case _ => None
+      }
+      val outers = bodies.zipWithIndex flatMap {
+        case (x:ModifierAtom,i) => {
+          val newIndexPrefix = s"${indexPrefix}${i}_"
+          x.modifier match {
+            case OuterModifier() => {
+              val from = generateFromClause(x.bodies, newIndexPrefix)
+              val joinCond = generateWhereClause(x.bodies, newIndexPrefix)
+              Some(s"${from} ON ${joinCond}")
             }
+            case _ => None
           }
         }
+        case _ => None
+      }
+      // full outer join
+      if (atoms isEmpty) {
+        outers.mkString(" FULL OUTER JOIN ")
+      } else {
+        (atoms.mkString(", ") +: outers).mkString(" LEFT OUTER JOIN ")
       }
     }
 
-    // resolve conditions
-    val conditionStr = z.conditions(0) map (compileCond(_, z))
+    val whereClause = generateWhereClause(z.bodies(0), "")
+    val whereClauseStr = if (whereClause.isEmpty) "" else "WHERE " + whereClause
+    val fromClauseStr = generateFromClause(z.bodies(0), "")
 
-    // handle group by
-    // map head terms, leaving out aggregation functions
-    val groupbyTerms = z.head.terms.zipWithIndex map { case (expr, index) =>
+    // check if an expression contains an aggregation function
+    def containsAggregation(expr: Expr) : Boolean = {
       expr match {
-        case FuncExpr(f, args, agg) => if (agg) None else Some("")
-        case _ => Some(compileExpr(expr, z, OriginalOnly, index, false))
+        case VarExpr(name) => false
+        case ConstExpr(value) => false
+        case FuncExpr(function, args, agg) => if (agg) agg else {
+          args.map(containsAggregation).foldLeft(false)(_ || _)
+        }
+        case BinaryOpExpr(lhs, op, rhs) => containsAggregation(lhs) || containsAggregation(rhs)
+        case TypecastExpr(lhs, rhs) => containsAggregation(lhs)
+        case Placeholder() => false
       }
     }
 
-    val groupbyStr = if (groupbyTerms.size == z.head.terms.size) {
+    // handle group by
+    // map head terms, leaving out aggregation functions
+    val groupbyTerms = z.head.terms.zipWithIndex flatMap { case (expr, index) =>
```

Comment:
Why do you `zipWithIndex`?  `parititon` or `filter` may be useful now since you nicely defined a boolean predicate.  You can chain that with a `map compileExpr`